File: backend/tests/students.test.py
from unittest.mock import patch
from fastapi.testclient import TestClient

from src.main import app

client = TestClient(app)


class TestGetStudents:
    def test_returns_all_students(self):
        with patch("src.routers.students.StudentModel.find_all") as mock:
            mock.return_value = [{"id": 1, "first_name": "Jean", "last_name": "Dupont", "class_id": 1}]
            res = client.get("/api/students/")
        assert res.status_code == 200
        assert len(res.json()) == 1

    def test_filters_by_class_id(self):
        with patch("src.routers.students.StudentModel.find_all") as mock:
            mock.return_value = []
            res = client.get("/api/students/?class_id=2")
        assert res.status_code == 200
        assert res.json() == []

    def test_filters_by_search_query(self):
        with patch("src.routers.students.StudentModel.find_all") as mock:
            mock.return_value = [{"id": 2, "first_name": "Marie", "last_name": "Curie", "class_id": 1}]
            res = client.get("/api/students/?q=Marie")
        assert res.status_code == 200
        assert res.json()[0]["first_name"] == "Marie"


class TestPostStudents:
    def test_creates_student_and_returns_201(self):
        with patch("src.routers.students.StudentModel.create") as mock:
            mock.return_value = {"id": 1, "first_name": "Jean", "last_name": "Dupont", "email": "jean@school.fr"}
            res = client.post("/api/students/", json={"first_name": "Jean", "last_name": "Dupont", "email": "jean@school.fr"})
        assert res.status_code == 201
        assert res.json()["first_name"] == "Jean"

    def test_returns_422_when_names_are_missing(self):
        res = client.post("/api/students/", json={"email": "jean@school.fr"})
        assert res.status_code == 422


class TestPutStudents:
    def test_updates_a_student(self):
        with patch("src.routers.students.StudentModel.update") as mock:
            mock.return_value = {"id": 1, "first_name": "Jean", "last_name": "Martin"}
            res = client.put("/api/students/1", json={"last_name": "Martin"})
        assert res.status_code == 200
        assert res.json()["last_name"] == "Martin"

    def test_returns_404_when_student_not_found(self):
        with patch("src.routers.students.StudentModel.update") as mock:
            mock.return_value = None
            res = client.put("/api/students/999", json={"last_name": "Unknown"})
        assert res.status_code == 404


class TestDeleteStudents:
    def test_deletes_and_returns_204(self):
        with patch("src.routers.students.StudentModel.delete") as mock:
            mock.return_value = True
            res = client.delete("/api/students/1")
        assert res.status_code == 204

    def test_returns_404_when_student_not_found(self):
        with patch("src.routers.students.StudentModel.delete") as mock:
            mock.return_value = False
            res = client.delete("/api/students/999")
        assert res.status_code == 404


class TestImportStudents:
    def test_returns_422_when_no_file_uploaded(self):
        res = client.post("/api/students/import")
        assert res.status_code == 422

    def test_imports_students_from_valid_csv(self):
        csv_content = b"first_name,last_name,email,class_label,year\nJean,Dupont,jean@school.fr,3A,2025"
        with patch("src.services.csv_service.ClassModel.find_all") as mock_classes, \
             patch("src.routers.students.StudentModel.bulk_create") as mock_create:
            mock_classes.return_value = [{"id": 1, "label": "3A", "year": 2025}]
            mock_create.return_value = [{"id": 1, "first_name": "Jean", "last_name": "Dupont"}]
            res = client.post(
                "/api/students/import",
                files={"file": ("students.csv", csv_content, "text/csv")},
            )
        assert res.status_code == 201
        assert res.json()["created"] >= 1

    def test_returns_errors_for_rows_with_missing_names(self):
        csv_content = b"first_name,last_name,email\n,Dupont,jean@school.fr\n,Curie,marie@school.fr"
        res = client.post(
            "/api/students/import",
            files={"file": ("students.csv", csv_content, "text/csv")},
        )
        assert res.status_code == 400
        assert "errors" in res.json()

File: backend/tests/students.test.js
const request = require('supertest');

jest.mock('../src/config/database', () => ({
  query: jest.fn(),
  connect: jest.fn(),
  on: jest.fn(),
}));

jest.mock('../src/services/imageService', () => ({
  resizePhoto: jest.fn().mockResolvedValue('/tmp/photo.jpg'),
}));

const app = require('../src/app');
const pool = require('../src/config/database');

describe('Students API', () => {
  beforeEach(() => jest.clearAllMocks());

  // ── GET /api/students ─────────────────────────────────
  describe('GET /api/students', () => {
    it('returns all students', async () => {
      pool.query.mockResolvedValue({
        rows: [{ id: 1, first_name: 'Jean', last_name: 'Dupont', class_id: 1 }],
      });

      const res = await request(app).get('/api/students');
      expect(res.status).toBe(200);
      expect(res.body).toHaveLength(1);
    });

    it('filters by class_id', async () => {
      pool.query.mockResolvedValue({ rows: [] });

      const res = await request(app).get('/api/students?class_id=2');
      expect(res.status).toBe(200);
      expect(res.body).toEqual([]);
    });

    it('filters by search query q', async () => {
      pool.query.mockResolvedValue({
        rows: [{ id: 2, first_name: 'Marie', last_name: 'Curie', class_id: 1 }],
      });

      const res = await request(app).get('/api/students?q=Marie');
      expect(res.status).toBe(200);
      expect(res.body[0].first_name).toBe('Marie');
    });
  });

  // ── POST /api/students ────────────────────────────────
  describe('POST /api/students', () => {
    it('creates a student and returns 201', async () => {
      pool.query.mockResolvedValue({
        rows: [{ id: 1, first_name: 'Jean', last_name: 'Dupont', email: 'jean@school.fr' }],
      });

      const res = await request(app)
        .post('/api/students')
        .send({ first_name: 'Jean', last_name: 'Dupont', email: 'jean@school.fr' });

      expect(res.status).toBe(201);
      expect(res.body.first_name).toBe('Jean');
    });

    it('returns 400 when names are missing', async () => {
      const res = await request(app)
        .post('/api/students')
        .send({ email: 'jean@school.fr' });

      expect(res.status).toBe(400);
    });
  });

  // ── PUT /api/students/:id ─────────────────────────────
  describe('PUT /api/students/:id', () => {
    it('updates a student', async () => {
      pool.query.mockResolvedValue({
        rows: [{ id: 1, first_name: 'Jean', last_name: 'Martin' }],
      });

      const res = await request(app)
        .put('/api/students/1')
        .send({ last_name: 'Martin' });

      expect(res.status).toBe(200);
      expect(res.body.last_name).toBe('Martin');
    });

    it('returns 404 when student not found', async () => {
      pool.query.mockResolvedValue({ rows: [] });

      const res = await request(app)
        .put('/api/students/999')
        .send({ last_name: 'Unknown' });

      expect(res.status).toBe(404);
    });
  });

  // ── DELETE /api/students/:id ──────────────────────────
  describe('DELETE /api/students/:id', () => {
    it('deletes and returns 204', async () => {
      pool.query.mockResolvedValue({ rowCount: 1 });

      const res = await request(app).delete('/api/students/1');
      expect(res.status).toBe(204);
    });

    it('returns 404 when student not found', async () => {
      pool.query.mockResolvedValue({ rowCount: 0 });

      const res = await request(app).delete('/api/students/999');
      expect(res.status).toBe(404);
    });
  });

  // ── POST /api/students/import ─────────────────────────
  describe('POST /api/students/import', () => {
    it('returns 400 when no file is uploaded', async () => {
      const res = await request(app).post('/api/students/import');
      expect(res.status).toBe(400);
    });

    it('imports students from a valid CSV (happy path)', async () => {
      // Class.findAll mock
      pool.query.mockResolvedValue({ rows: [{ id: 1, label: '3A', year: 2025 }] });

      // bulkCreate: mock connect → transaction
      const mockClient = {
        query: jest.fn()
          .mockResolvedValueOnce({}) // BEGIN
          .mockResolvedValueOnce({ rows: [{ id: 1, first_name: 'Jean', last_name: 'Dupont' }] }) // INSERT
          .mockResolvedValueOnce({}), // COMMIT
        release: jest.fn(),
      };
      pool.connect.mockResolvedValue(mockClient);

      const csv = 'first_name,last_name,email,class_label,year\nJean,Dupont,jean@school.fr,3A,2025';

      const res = await request(app)
        .post('/api/students/import')
        .attach('file', Buffer.from(csv), { filename: 'students.csv', contentType: 'text/csv' });

      expect(res.status).toBe(201);
      expect(res.body.created).toBeGreaterThan(0);
    });

    it('returns errors for rows with missing names', async () => {
      pool.query.mockResolvedValue({ rows: [] });
      const mockClient = { query: jest.fn().mockResolvedValue({}), release: jest.fn() };
      pool.connect.mockResolvedValue(mockClient);

      // All rows have empty first_name
      const csv = 'first_name,last_name,email\n,Dupont,jean@school.fr\n,Curie,marie@school.fr';

      const res = await request(app)
        .post('/api/students/import')
        .attach('file', Buffer.from(csv), { filename: 'students.csv', contentType: 'text/csv' });

      expect(res.status).toBe(400);
      expect(res.body.errors).toBeDefined();
    });
  });
});
